rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+---
+feature: system-parameters
+authors:
+  - "Eric Fu"
+start_date: "2023/01/11"
+---
+
+# RFC: System Parameters
+
+## Motivation
+
+Currently, we have 3 kinds of configurations:
+
+- Command-line arguments
+- Configuration file `risingwave.toml`
+- Session parameters
+
+These can not meet the requirements for several reasons.
+
+1. **Consistency**. `risingwave.toml` is a local file, which makes it possible that multiple nodes could see different parameters and this is absolutely unacceptable. For example, to configure the remote directory for storing data, users may use the `data_directory` parameter. If some nodes use `data_directory = 'hummock_001'` and others use `data_directory = 'hummock_002'`, the stored data will be corrupted.
+2. **Mutability**. Users are unable to `SET` a config after the process started. The root cause is still the local file - changes of configs must be persisted in some cluster-level storage.
+
+This RFC proposes to introduce centralized configurations on Meta Service and data will be persisted in etcd.
+
+For the sake of clarity, in this doc, we will use the term **"parameters"** for system and session parameters, and **"config"** for command-line arguments or config files.
+
+**Note:** Previously I think this might not be an urgent issue until I realized that our Cloud platform couples tightly with more and more configurations. The later we commence, the heavier burden it will become. 
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Parameters (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Parameters
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
+
+Here we could borrow the syntax from Postgres:
+
+```sql
+-- Set or reset a parameter
+ALTER SYSTEM / DATABASE SET parameter TO value / 'value' / DEFAULT;
+
+-- List all parameters keys & values
+SHOW PARAMETERS;
+```
+
+> **Note**: PG uses `SHOW ALL` because it doesn't has other `SHOW` commands like us. Which one is better?
+
+
+References:
+
+- [PostgreSQL: Documentation: 15: ALTER SYSTEM](https://www.postgresql.org/docs/current/sql-altersystem.html)
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+- [PostgreSQL: Documentation: 15: SHOW](https://www.postgresql.org/docs/current/sql-show.html)
+
+Similar to other system metadata e.g. catalog,
+
+- Changes to global parameters should be propagated to all nodes, in the cluster in a deterministic order ideally
+- When a node reboots or recovers, the latest snapshot of parameters should be passed from Meta Service.
```

Comment:
Could you please elaborate it more about the initialization of the "snapshot" (in meta store)? For example,
- Do we still have configuration file for the cluster? If so, do we only persist the configuration on the first run, or do we also detect changes in the configuration file on restarts?
- Do we persist the default values of the entries that do not appear in `.toml`? We should be cautious about some default values that are "computed" instead of constants, as persisting them may be confusing.

Reply:
Just to avoid misunderstanding: the "snapshot" here refers to the current system parameters in Meta Service at the beginning of subscription, just like the "snapshot" of the catalog or Hummock versions.

Answers to questions:

> Do we still have configuration file for the cluster? If so, do we only persist the configuration on the first run, or do we also detect changes in the configuration file on restarts?

Basically not. It may be useful for a few cases like https://github.com/risingwavelabs/rfcs/pull/34#issuecomment-1381424149, but generally, I don't think we need such a design.

> Do we persist the default values of the entries that do not appear in `.toml`? We should be cautious about some default values that are "computed" instead of constants, as persisting them may be confusing.

Computed values (like parallelism described in https://github.com/risingwavelabs/rfcs/pull/34#discussion_r1069061515) should be stored as `'auto'` or `0`, so does this solves your concern? In this way, either persisting default values or not can work well, but I still prefer not to persist default values because it allows us to tune the defaults for better performance or whatever.

Reply:
> Basically not. It may be useful for a few cases like [#34 (comment)](https://github.com/risingwavelabs/rfcs/pull/34#issuecomment-1381424149), but generally, I don't think we need such a design.

Does it mean that we need to manually initialize this on the first run, before issuing any streaming jobs? It sounds not that friendly for development and playground purposes.

> but I still prefer not to persist default values

Sounds like a reasonable choice! But we should still be cautious when upgrading if some users rely on the previous default values.

Reply:
> Does it mean that we need to manually initialize this on the first run, before issuing any streaming jobs? It sounds not that friendly for development and playground purposes.

IIUC, we use minio for development and playground purposes now, and that's why we can use a default value. We can keep this behavior unchanged. 

But for productional deployment or cloud platform, a unique S3 bucket must be manually specified for sure.

Reply:
Not only the `data_directory`, but also the `state_store_url` which tells which state store backend we should use. 🤔

Reply:
I don't think the `data_directory` will be a big problem as long as it's not mutable and the deployment tool enforces the value. What really matters is how we keep the caches on various nodes updated and how to make sure things do not break when parameter change happens.